### PR TITLE
osd: avoid transaction append much of the time

### PR DESCRIPTION
**Pull request by liewegas**
I think this will reclaim most of the hit we took for copying the ops in Transaction::append()

### CODE REVIEW

**Comment by liewegas** [comment]
@yuandong1222 can you review?  this passed a sanity test run

File: src/messages/MOSDECSubOpWrite.h [hunk 1]
```diff
@@ -26,30 +26,35 @@ class MOSDECSubOpWrite : public Message {
...
-  MOSDECSubOpWrite(ECSubWrite &op)
-  : Message(MSG_OSD_EC_WRITE, HEAD_VERSION, COMPAT_VERSION),
-    op(op) {}
+    : Message(MSG_OSD_EC_WRITE, HEAD_VERSION, COMPAT_VERSION), op(NULL)
+  {}
+  MOSDECSubOpWrite(ECSubWrite *op)
+    : Message(MSG_OSD_EC_WRITE, HEAD_VERSION, COMPAT_VERSION), op(op)
+  {}
+  ~MOSDECSubOpWrite() {
+    delete op;
```
Comment (yuandong1222) [inline src/messages/MOSDECSubOpWrite.h]: Maybe we should make sure op == NULL when deleting.

Comment (yuandong1222) [inline src/messages/MOSDECSubOpWrite.h]: Forget it, delete will do it.

**Comment by yuandong1222** [comment]
Looks Good To Me.

**Comment by yuandong1222** [comment]
Should I see this?

On 30 January 2015 at 09:26, Automated make check for Ceph <
notifications@github.com> wrote:

> FAIL: the output of run-make-check.sh on centos-centos7 for 313ec6b
> <https://github.com/ceph/ceph/commit/313ec6bb576ba1d9290e2102c5d10a56ab6d1479>
> is http://paste2.org/_nLe5FaLd
>
> [image: :octocat:] *Sent from GH <http://nodegh.io>.*
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ceph/ceph/pull/3478#issuecomment-72138345>.
>



-- 
Dong Yuan
Email:yuandong1222@gmail.com

**Comment by dachary** [comment]
@yuandong1222 the output of the run-make-check.sh script (which you can run on your own machine) shows what seems to be a compilation error:
<pre>
 CXXLD ceph_test_rados_api_nlist
In file included from test/encoding/types.h:89:0,
from test/encoding/ceph_dencoder.cc:18:
test/encoding/ceph_dencoder.cc: In instantiation of 'void DencoderImplNoFeature<T>::copy() [with T = ECSubWrite]':
test/encoding/ceph_dencoder.cc:415:1: required from here
./osd/ECMsgTypes.h:66:21: error: 'const ECSubWrite& ECSubWrite::operator=(const ECSubWrite&)' is private
const ECSubWrite& operator=(const ECSubWrite& other);
^
test/encoding/ceph_dencoder.cc:139:8: error: within this context
*n = *this->m_object;
^
In file included from test/encoding/types.h:89:0,
</pre>

**Comment by yuandong1222** [comment]
I can make ceph_test_rados_api_nlist in my environment, but make check hangs.

I try on both CentOS and  Ubuntu. 

On CentOS the last log is:
[ RUN      ] TestImageWatcher.RequestLockPostTryLockFailed
SKIPPING
[       OK ] TestImageWatcher.RequestLockPostTryLockFailed (0 ms)
[ RUN      ] TestImageWatcher.NotifyHeaderUpdate
SKIPPING
[       OK ] TestImageWatcher.NotifyHeaderUpdate (0 ms)
[----------] 15 tests from TestImageWatcher (1 ms total)

[----------] 33 tests from TestLibRBD
seed 29960
[ RUN      ] TestLibRBD.CreateAndStat
using old format


While on Ubuntu, the last log is 
PASS: unittest_mon_moncap
PASS: unittest_mon_pgmap
PASS: unittest_ipaddr
PASS: unittest_texttable
PASS: unittest_on_exit
PASS: unittest_rbd_replay
PASS: unittest_readahead
PASS: unittest_tableformatter

Did I miss something?

**Comment by athanatos** [comment]
I've commented on the last patch.  I've got some serious reservations about that route from a maintenance standpoint.  Re-add needs-qa once you've addressed/disagreed with my comments and rebased so it merges cleanly.

**Comment by athanatos** [comment]
I'm testing a different version of this branch rebased on master.

**Comment by athanatos** [comment]
How about https://github.com/athanatos/ceph/tree/wip-txn-noappend-sam-rebased instead?

**Comment by liewegas** [comment]
yeah, that version is better

**Pull request closed without merging**